src/v/coproc/materialized_topics_frontend.h

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "cluster/types.h"
+#include "coproc/exception.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/sharded.hh>
+#include <seastar/core/shared_future.hh>
+#include <seastar/core/shared_ptr.hh>
+
+#include <absl/container/node_hash_map.h>
+
+namespace coproc {
+
+/// \brief Thrown when the cluster subsystem fails to replicate the materialized
+/// topic.
+class materialized_topic_replication_exception final : public exception {
+    using exception::exception;
+};
+
+/// This class is used to debounce requests to make materialzed topics. A single
+/// coprocessor will (most likely) make multiple requests to create the same
+/// materialized topic across shards as long at it has an input topic with more
+/// then 1 partition. Fibers serving these partitions can use this class to
+/// reconcile and debounce duplicate requests to make a materialized topic.
+class materialized_topics_frontend {
+public:
+    /// \brief Mapping between requested topic_namespace to create and shared
+    /// promise that is associated with an outstanding request
+    using underlying_t = absl::node_hash_map<
+      model::topic_namespace,
+      ss::lw_shared_ptr<ss::shared_promise<>>>;
```

Comment:
let's avoid shared_promise if possible. it's very hard to use correctly and we often trace bugs to its usage.

Reply:
hmm its the right abstraction here :/ I can work around that though if there are bugs and stuff sure

Reply:
why is a shared_promise necessary? generally i think it is fair to say that each time we've used shared_promise we've had to fix a bug with its usage because it doesn't behave like a normal promise.

Reply:
Its not, its been removed as of the latest force-push